### DISCUSSION TITLE
Plans to support WebGPU?

### DISCUSSION THREAD

**Discussion by zro00** [pixiv/three-vrm#1278] [General]
Are there any plans to use start using the new webgpu api with this library?
Currently trying THREE.WebGPURenderer returns an error.

**Comment by 0b5vr** [comment]
There is no plan to support WebGPU at the moment.
Do you have any reason to use WebGPU instead of WebGL? Asking because we are simply interested what is the advantage to use WebGPU.

**Comment by zro00** [comment]
Scenes with a ton of models and texture hit performance limitation pretty quickly
WebGPU is much more performant overall this will immediately unlock critical performance gain without having the developer resort to adding time consuming performance hacks (off canvas rendering, etc ..). It would definitely enable much more complex scenes with more particles, physics, larger terrain, etc ...

One major win as well is that both Google and Apple are on board with WebGPU. This means that mobile devices will stand to gain the most from the free performance boost.

WebGPU is the successor to WebGL and both threejs and babylonjs are adapting it natively. 
WebGPU has more momentum and better feature set then WebGL did in its entire lifetime. WASM and AI applications are both now starting to leverage that API.

**Comment by zro00** [comment]
1-2 years from now everyone will be using or want to use WebGPU. If i want to use WebGPU then I can't use this library for threejs which is a shame. 
The only other web native port of the VRM spec was for babylonjs and that repo hasn't seen active development in quite a while.

**Comment by 0b5vr** [comment]
Thank you for the details!

I'm sure we have to support WebGPU as the demand grows, because the performance is a critical factor for most of 3D products as you said.

One bottleneck is that we are using a custom shader for MToon material, and it needs a WebGL-WebGPU cross support shader for the material.
Recently, sunag is working on a nobel method to write cross support shader.

https://threejs.org/examples/#webgpu_tsl_editor

We are yet to find our time to try this, but I'm personally very interested in this at least 😛

**Reply by zro00** [reply]
@0b5vr There is a rust utility created by some key members of webGPU working group that helps with the translation process:
https://github.com/gfx-rs/naga

Here a slide deck (rather out of date at this point) talking about some of the high level differences:
https://www.khronos.org/assets/uploads/developers/presentations/Jasper-St.-Pierre---Porting-an-existing-WebGL-application-to-WebGPU_.pdf

**Reply by 0b5vr** [reply]
Three.js GLSL uses a lot of "ShaderChunks" to make it connect with Three.js pipelines and so do the MToon. I think the codebase is pretty difficult to convert using third party libraries.

See, The MToon material:
https://github.com/pixiv/three-vrm/blob/dev/packages/three-vrm-materials-mtoon/src/shaders/mtoon.frag

See, The tool I usually use to deal with the ShaderChunk shenanigans. You can grasp how sphagettiful the system is:
https://three-phong-shader-inspector.glitch.me/

**Comment by ButzYung** [comment]
IMO whether three-vrm can use WebGPU or not is more on three.js itself. three.js's support on WebGPU is still quite basic and I don't think it provides enough features to render VRM in full specs at this moment.

Besides, to think everyone will be using WebGPU in 1-2 years is a bit optimistic. Right now, the only major browser that supports WebGPU is Chrome, and its support is not fully cross platform yet. Safari doesn't even have ETA, and for historical reasons I don't hold too much hope to expect it to come soon.
